code/segmentedtickerplant/stplog.q

```diff
@@ -0,0 +1,177 @@
+// Utilites for periodic tp logging in stp process
+
+// Live logs and handles to logs for each table
+currlog:([tbl:`symbol$()]logname:`symbol$();handle:`int$())
+
+// View of log file handles for faster lookups
+loghandles::exec tbl!handle from currlog
+
+\d .stplg
+
+// Create stp log directory
+// Log structure `:stplogs/date/tabname_time
+createdld:{[name;date]
+  $[count dir:getenv[`KDBSTPLOG];
+    [.os.md dir;.os.md dldir::hsym`$raze dir,"/",string name,date];
+    [.lg.e[`stp;"log directory not defined"];exit]
+  ]
+ };
+
+// Functions to generate log names in one of five modes ///////////////////////////////
+
+logname:enlist[`]!enlist ()
+
+// Default stp mode is tabperiod
+// TP log rolled periodically (default 1 hr), 1 log per table
+logname[`tabperiod]:{[dir;tab;p]
+  ` sv(hsym dir;`$string[tab],raze[string"du"$p]except".:")
+ };
+
+// Standard TP mode - write all tables to single log, roll daily
+logname[`none]:{[dir;tab;p]
+  ` sv(hsym .stplg.dldir;`$string[.proc.procname],"_",string[.z.d]except".")
+ };
+
+// Periodic-only mode - write all tables to single log, roll periodically intraday
+logname[`periodic]:{[dir;tab;p]
+  ` sv(hsym dir;`$"periodic",raze[string"du"$p]except".:")
+ };
+
+// Tabular-only mode - write tables to separate logs, roll daily
+logname[`tabular]:{[dir;tab;p]
+  ` sv(hsym dir;`$string[tab],"_",string[.z.d]except".")
+ };
+
+// Custom mode - mixed periodic/tabular mode
+// Tables are defined as periodic, tabular, tabperiod or none in config file stpcustom.csv
+// Tables not specified in csv are not logged
+logname[`custom]:{[dir;tab;p]
+  logname[custommode tab][dir;tab;p]
+ };
+
+///////////////////////////////////////////////////////////////////////////////////////
+
+// Update and timer functions in three batch modes ////////////////////////////////////
+
+upd:zts:enlist[`]!enlist ()
+
+// If set to autobatch, publish and write to disk will be run in batches
+upd[`autobatch]:{[t;x]
+  .stpps.upd[t;x];
+ };
+
+zts[`autobatch]:{
+  {[t;x] .stpps.pub[t;x];
+  `..loghandles[t] enlist(`upd;t;x)}'[.stpps.t;value each .stpps.t];
+  @[`.;.stpps.t;@[;`sym;`g#]0#];
+  ts .z.p;
```

Comment:
Should this definitely be .z.p passed in here? It is generally best to use .z.p and then checked based on that if possible as the OS will return .z.p faster than .z.P (maybe 4 times faster)